### PR TITLE
CUMULUS-2323 - Update S3 provider to take optional granule.file.bucket parameter

### PR DESCRIPTION
**Pull request by Jkovarik**
Draft PR - for illustrative purposes only (for now)

### CODE REVIEW

**Comment by ngachung** [comment]
@Jkovarik Is there an update on when this change will be reviewed and delivered? We have some issues that are blocked by this. Thank you!

File: tasks/sync-granule/GranuleFetcher.js [hunk 1]
```diff
@@ -237,25 +237,6 @@ class GranuleFetcher {
...
-  /**
-   * Add a bucket property to the given file
-   *
-   * Note: This returns a copy of the file parameter, it does not modify it.
-   *
-   * @param {Object} file - an object containing a "name" property
-   * @returns {Object} the file with a bucket property set
-   * @throws {Error} if the file didn't have a matching config
-   */
-  addBucketToFile(file) {
```
Comment (Jkovarik) [inline tasks/sync-granule/GranuleFetcher.js:249]: This isn't being used anywhere.  Removed and removed relevant tests.

File: packages/ingest/src/FtpProviderClient.ts [hunk 2]
```diff
@@ -200,20 +200,21 @@ class FtpProviderClient {
...
    *    the S3 URI and ETag of the destination file
    */
-  async sync(
-    remotePath: string,
-    bucket: string,
-    key: string
-  ): Promise<{s3uri: string, etag: string}> {
-    const remoteUrl = `ftp://${this.host}/${remotePath}`;
-    const s3uri = S3.buildS3Uri(bucket, key);
+  async sync(params: {
```
Comment (Jkovarik) [inline packages/ingest/src/FtpProviderClient.ts:203]: Changes made to this class are entirely related to the parameterization of the .sync ProviderClient base class/this class  and/or refactors related to param names/etc.

File: packages/ingest/src/FtpProviderClient.ts [hunk 3]
```diff
@@ -6,7 +6,11 @@ import Logger from '@cumulus/logger';
 import { Socket } from 'net';
 import { recursion } from './recursion';
 import { lookupMimeType, decrypt } from './util';
-import { FtpProviderClientListItem, ProviderClientListItem } from './types';
+import {
```
Comment (Jkovarik) [inline packages/ingest/src/FtpProviderClient.ts:9]: Changes made to this class are entirely related to the parameterization of the .sync ProviderClient base class/this class and/or refactors related to param names/etc.

File: packages/ingest/src/S3ProviderClient.ts [hunk 4]
```diff
@@ -61,22 +61,28 @@ class S3ProviderClient implements ProviderClient {
   /**
```
Comment (Jkovarik) [inline packages/ingest/src/S3ProviderClient.ts:61]: Updates to this method are the 'guts' of the change, we're adding a bucket parameter to this expected params for this sync method to allow for an override to the Provider bucket value (defined by the provider configuration) on a per-file basis

File: packages/ingest/src/SftpProviderClient.js [hunk 5]
```diff
@@ -144,7 +144,10 @@ class SftpProviderClient {
    * @returns {Promise.<{ s3uri: string, etag: string }>} an object containing
    *    the S3 URI and ETag of the destination file
    */
-  async sync(remotePath, bucket, key) {
+  async sync(params) {
```
Comment (Jkovarik) [inline packages/ingest/src/SftpProviderClient.js:147]: Making the signature of this (non inherited class) match ProviderClient.   Given time I might put in a seperate PR to typescript this.

File: packages/ingest/src/granule.ts [hunk 6]
```diff
@@ -199,8 +199,15 @@ export async function handleDuplicateFile(params: {
   target: { Bucket: string, Key: string },
   duplicateHandling: DuplicateHandling,
   checksumFunction?: (bucket: string, key: string) => Promise<[string, string]>,
-  syncFileFunction?: (bucket: string, key: string) => Promise<void>,
-  ACL?: string
+  syncFileFunction?: (params: {
```
Comment (Jkovarik) [inline packages/ingest/src/granule.ts:202]: Updates syncFileFunction to allow for the optional bucket parameter.  Adds optional sourceBucket key to params.

File: packages/ingest/test/test-HttpsProviderClient.js [hunk 7]
```diff
@@ -129,26 +129,26 @@ test('download() downloads a file', async (t) => {
 });
 
 test('sync() downloads remote file to s3 with correct content-type', async (t) => {
-  const bucket = randomString();
-  const key = 'syncedFile.json';
+  const destinationBucket = randomString();
```
Comment (Jkovarik) [inline packages/ingest/test/test-HttpsProviderClient.js:132]: These changes all just are updating this test to use sync's parameterized call, and updating the variables from generics like `bucket` and `key` to be more consistent.

File: packages/ingest/test/test-granule.js [hunk 8]
```diff
@@ -339,3 +342,161 @@ test('unversionFilename returns filename without version stamp if present', (t)
 
   t.is(expected, actual);
 });
+
+test('handleDuplicateFile throws DuplicateFile if method is called and duplicateHandling is set to "error"', async (t) => {
```
Comment (Jkovarik) [inline packages/ingest/test/test-granule.js:346]: This method wasn't tested at all, so I added tests that cover the logic branches.   Given what that looks like, there's probably room for a refactor to handleDuplicateFile/associated logic, however that's best saved for a PR and *probably* out of scope for this ticket.

File: example/cumulus-tf/iam.tf [hunk 9]
```diff
@@ -1,11 +1,18 @@
 data "aws_cloudformation_export" "fake_s3_provider" {
   name = "cumulus-fake-s3-provider-bucket"
 }
+data "aws_cloudformation_export" "fake_s3_provider_alternate" {
```
Comment (Jkovarik) [inline example/cumulus-tf/iam.tf:4]: The updates to this file are all required to setup an alternate bucket for ingest testing.

File: example/fake-provider-cf.yml [hunk 10]
```diff
@@ -23,6 +23,10 @@ Outputs:
     Value: !Ref S3ProviderBucket
     Export:
       Name: cumulus-fake-s3-provider-bucket
+  S3AltProviderBucket:
```
Comment (Jkovarik) [inline example/fake-provider-cf.yml:26]: Updates here are adding the alternate bucket to the CF stack

File: example/spec/parallel/SyncGranuleTask/TestAlternateFileBucketSpec.js [hunk 11]
```diff
@@ -0,0 +1,155 @@
+
+'use strict';
```
Comment (Jkovarik) [inline example/spec/parallel/SyncGranuleTask/TestAlternateFileBucketSpec.js:2]: This is a component integration test that runs sync granule with a payload like what we're expecting to work for this ticket.

File: example/spec/parallel/SyncGranuleTask/TestAlternateFileBucketSpec.js [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+const { deleteCollection } = require('@cumulus/api-client/collections');
+const { deleteProvider } = require('@cumulus/api-client/providers');
+const { lambda } = require('@cumulus/aws-client/services');
+const S3 = require('@cumulus/aws-client/S3');
+
+const { loadConfig } = require('../../helpers/testUtils');
+const { fetchFakeS3ProviderBuckets } = require('../../helpers/Providers');
+
+describe('The SyncGranule task with a granule file using an alternate bucket', () => {
```
Comment (Jkovarik) [inline example/spec/parallel/SyncGranuleTask/TestAlternateFileBucketSpec.js:21]: Consider doc/reword this test

File: packages/ingest/test/test-granule.js (hunk 8 above)
Comment (nemreid) [reply packages/ingest/test/test-granule.js:346]: > ... and ~_probably_~ *_definitely_* out of scope for this ticket.

File: packages/ingest/test/test-granule.js [hunk 13]
```diff
@@ -339,3 +342,161 @@ test('unversionFilename returns filename without version stamp if present', (t)
...
+
+test('handleDuplicateFile returns an empty array if duplicateHandling is set to "skip"', async (t) => {
+  const actual = await handleDuplicateFile({
+    duplicateHandling: 'skip',
+  });
+  const expected = [];
+  t.deepEqual(actual, expected);
+});
+
+test('handleDuplicateFile calls syncFileFunction with expected arguments if duplicateHandling is set to "version" and syncFileFunction is provided', async (t) => {
```
Comment (nemreid) [inline packages/ingest/test/test-granule.js:364]: I'm not convinced we need it, but there is no test for `duplicateHandling is set to "version" and SyncFileFunction is _not_ provided`?

**Review by nemreid — APPROVED** [review]
Great work! I have a few small comments, but nothing that I consider blocking at this time.

File: tasks/sync-granule/GranuleFetcher.js [hunk 14]
```diff
@@ -360,8 +342,7 @@ class GranuleFetcher {
 
       // bind arguments to sync function
       const syncFileFunction = providerClient.sync.bind(
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply tasks/sync-granule/GranuleFetcher.js:344]: Actually, fun story, that's not quite as trivially true as we thought when chatting

Bind:

`The bind() method creates a new function that, when called, has its this keyword set to the provided value, with a given sequence of arguments preceding any provided when the new function is called.`

> has its this keyword set to the provided value< being the important bit.   Not setting this actually breaks the unit tests, and I believe functionality, as passing *sync* in doesn't retain `this` for the various providers. 

A refactor here would require an update to `handleDuplicateFile`'s signature.  I may take a stab at that, but would prefer it to be a seperate PR.

**Comment by Jkovarik** [comment]
@nemreid  one further commit to review.   On re-review, it turns out the 'bucket' file.granule value is being utilized by discover granules to populate collection config outputs per-file.    I could in *theory* remove the bucket output from DiscoGranules and *our* integration tests would pass, however there's clearly a user contract there, and while *sync granules* updates this bucket value to be the destination bucket, other upstream components have different expectations and may not be using sync granules. 

As such, I opted to go with the second option for this and add a optional field to the *sync granules* granule.file object.       Take a look and let me know what you think.     

407bd6719

File: example/spec/parallel/SyncGranuleTask/TestAlternateFileBucketSpec.js (hunk 12 above)
Comment (Jkovarik) [reply example/spec/parallel/SyncGranuleTask/TestAlternateFileBucketSpec.js:24]: f2f80eb62

File: packages/ingest/test/test-granule.js (hunk 13 above)
Comment (Jkovarik) [reply packages/ingest/test/test-granule.js:364]: Actually, yes.  This needs a short refactor, as it's not checking the s3Object.moveObject call.    Good catch.

Comment (Jkovarik) [reply packages/ingest/test/test-granule.js:385]: d1f67afb9

**Pull request merged**